start epics variable dump tool

rlcee opened the pull request:
Creates a class EpicsTool to retrieve epics variables from the offline replica of the online epics database.  The intent is bring epics info to the offline user in a simple, familiar format. This class could be used to fill ntuples, for example.  A new bin, epicsTool, is built which can dump epics variables at the command line.  Features can be added as needed.  Also, more sophisticated queries could be needed if it starts running too slowly due to large datasets.

FNALbuild commented:
Hi @rlcee,
You have proposed changes to files in these packages:
- DbService

which require these tests: build.

@Mu2e/fnalbuild-users, @Mu2e/write have access to CI actions on [main](/Mu2e/Offline/tree/main).





:hourglass: The following tests have been triggered for a7fea903cf494550743b01cca1d96434f0161a6e: build  (Build queue is empty)


[About FNALbuild.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#GitHub_Pull_Request_Procedures_and_FNALbuild) [Code review on Mu2e/Offline.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#Code_Review)

kutschke commented:
@gaponenko are you available to do a review of this PR?

FNALbuild commented:
:sunny: The build tests passed at a7fea903cf494550743b01cca1d96434f0161a6e.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged a7fea903cf494550743b01cca1d96434f0161a6e at 67075dc0e135cffbd66e576ded5242f251b78808 |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/scons.log). Build time: 12 min 15 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :white_check_mark: | [TODO (0) FIXME (0) in 5 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/artifact/clang-tidy.log) |


N.B. These results were obtained from a build of this Pull Request at a7fea903cf494550743b01cca1d96434f0161a6e after being merged into the base branch at 67075dc0e135cffbd66e576ded5242f251b78808.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1288/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

gaponenko commented on DbService/src/epicsTool_main.cc:
```diff
@@ -0,0 +1,105 @@
+  std::string word, var, time;
+  size_t i = 1;
+  size_t lim(argc);
+  while (i < lim) {
```
Please consider using boost::program_options here for more maintainable code.

rlcee replied:
@kutschke @brownd1978 do we have a policy?  I've gone around and around on this (and other similar things) - if we declare a standard, I'll just follow it.

kutschke replied:
I don't think that we have ever discussed a policy - we have so few main programs.   I have also written my own command line parser in at least one of the mains. I think it would be good to move to a standard command line parser too as targets of opportunity.  

Are there reasonably options other than boost::program_options  .  I will ask Kyle if anything is in the standard, now or upcoming.

gaponenko replied:
Thank you for updating  the code to use boost::program_options

gaponenko reviewed (changes requested):
Hello,   I made a number of inline comments.  Overall, I would like to improve the  API for EpicsVar.  Ideally it should be constructed in a self consistent state and should not need modifier methods.   Also using "now" as a time reference in EpicsTool does not seem to be consistent with Offline philosophy.

rlcee commented on DbService/src/EpicsTool.cc:
```diff
@@ -0,0 +1,198 @@
+    while (i > 0 && ev[i].ttime() > tt) {
+      i--;
+    }
+    evt.push_back(ev[i]);
```
This seems to work. The loop stops when ev[i].ttime() < tt, then it reports ev[i].ttime()

gaponenko replied:
OK, I got it.   Thanks.

rlcee commented on DbService/inc/EpicsVar.hh:
```diff
@@ -0,0 +1,63 @@
+  int8_t _nanosecs;
+  int8_t _severity_id;
+  int8_t _status_id;
+  int _num_val;
+  double _float_val;
+  std::string _str_val;
```
I was simply reflecting the epics table content.  I can add the variant, but I'm not sure how I feel about removing the other accessors.  It feels a bit presumptive - I'm not sure of the rules of how these are filled.  I can ask.

gaponenko replied:
I think a EpicsVar contains a value of only a single type at a time - not an int and a float and a string simultaneously? Then it is neater to have a single member to represent this, the variant.    More importantly,  std::variant would provide a safer way to use the content - it throws if one tries to read  a wrong type from it.    The current design does not protect the user from reading garbage.   If I am wrong and EpicsVar is intended to contain all 3 types at once, please say so.

brownd1978 commented:
I use getopt_long in KinKal and TrackToy test programs.  It's a bit clumsy
but does what's expected and is standard.   My strategy is to make the
program a function below main itself, then I can have multiple mains with
the same infrastructure but different default arguments.

On Wed, May 11, 2022 at 9:34 AM Ray Culbertson ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In DbService/inc/EpicsVar.hh
> <https://github.com/Mu2e/Offline/pull/787#discussion_r870528474>:
>
> > +  int _num_val;
> +  double _float_val;
> +  std::string _str_val;
>
> I was simply reflecting the epics table content. I can add the variant,
> but I'm not sure how I feel about removing the other accessors. It feels a
> bit presumptive - I'm not sure of the rules of how these are filled. I can
> ask.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/787#discussion_r870528474>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH572A3X7347FLO2X7FRTVJPOSNANCNFSM5VJJGZMQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
David Brown            ***@***.***
Office Phone (510) 486-7261
Lawrence Berkeley National Lab
M/S 50R5008 (50-6026C)  Berkeley, CA 94720

FNALbuild commented:
:memo: The HEAD of `main` has changed to fd90897e01ac1aa1dd2d81bbef7f2f7045b66294. Tests are now out of date.

gaponenko commented:
Hi Dave,

getopt_long is also a well established option.  The upside is that it is 
in glibc and does not need boost.  The downside is it is not C++ but C.
A  C++ oriented package is more convenient when used with C++.

Andrei

On Wed, 11 May 2022, David Brown wrote:

> I use getopt_long in KinKal and TrackToy test programs.  It's a bit clumsy
> but does what's expected and is standard.

rlcee commented on DbService/src/EpicsVar.cc:
```diff
@@ -0,0 +1,58 @@
+  _csv = csv;
+
+  std::vector<std::string> sv;
+
```
I added a csv example.  The documents that exist are out of date on this particular point.

rlcee commented on DbService/src/EpicsVar.cc:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  _channel_id = std::stoll(sv[0]);
+  _smpl_time_s = sv[1];
```
added

rlcee commented:
I fixed what I could, resolved the small fixed comments.  Waiting on final word on the getopts.  Waiting to hear the proper way to fill the variant.

kutschke commented:
> Hi Dave, getopt_long is also a well established option. The upside is that it is in glibc and does not need boost. The downside is it is not C++ but C. A C++ oriented package is more convenient when used with C++. Andrei
> […](#)
> On Wed, 11 May 2022, David Brown wrote: I use getopt_long in KinKal and TrackToy test programs. It's a bit clumsy but does what's expected and is standard.

I don't consider the dependence on boost to be a downside.  I think that having a native C++ interface is preferred over a C interface - type safety and memory management are better.

Here is what Kyle had to say:

> Boost program options is a reasonable choice--art has used it for a long time in processing program options for its executables (art, product_sizes_dumper, etc.).  Some downsides with Boost program options is that uses somewhat abstruse syntax.  It also doesn't support subcommands (e.g. git diff <committish>, where diff is the subcommand). That has not been a problem for us so far.  On the positive side, it is a well-established library and unlikely to go away soon.
> 
> There are several alternatives to Boost--but I'm not aware of anything that has been proposed for the standard, probably because it's not a general-enough utility. I'm using CLI11 (https://cliutils.github.io/CLI11/book/) in my LDRD project (it supports subcommands and has better syntax than Boost), but its support base is much smaller and I'm not sure it's worth it.
> 
> My $0.02,
> Kyle

I don't know boost program options.  I recommend that Ray have a quick look at it.  If the syntax is not too weird, let's go with it for Mu2e main programs and fix them as targets of opportunity.  It's Dave's call for what he does in KTrack etc.

My reasoning is to change to the more powerful tool while the cost of changing is small.

rlcee commented:
boost program options was straightforward with a good example.

kutschke commented:
@gaponenko has Ray answered your questions?  If so, please approve.  Otherwise follow-up.

kutschke commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 69b7912c50bb16f10531cba6b0c1ef547130b643: build  (Build queue is empty)

FNALbuild commented:
:sunny: The build tests passed at 69b7912c50bb16f10531cba6b0c1ef547130b643.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged 69b7912c50bb16f10531cba6b0c1ef547130b643 at 39be3ce878b7d3e1d9c3aa786fc8578053da9a8a |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/scons.log). Build time: 18 min 21 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :white_check_mark: | [TODO (0) FIXME (0) in 7 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/artifact/clang-tidy.log) |


N.B. These results were obtained from a build of this Pull Request at 69b7912c50bb16f10531cba6b0c1ef547130b643 after being merged into the base branch at 39be3ce878b7d3e1d9c3aa786fc8578053da9a8a.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1300/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

gaponenko reviewed (changes requested):
Hi Ray,  thank you for making the updates.   I have resolved relevant conversations in github.   Some threads are still open, let's follow up on those:
- EpicsVar design exposing "garbage" values vs the use of std::variant
- The comment about "EpicsVar constructor should set the object state in a consistent way"
- The use of "now" as a time reference that I think is contrary to Offline philosopy - what do @kutschke and @brownd1978 think about  this?

rlcee commented:
> * EpicsVar design exposing "garbage" values vs the use of std::variant

I think (not sure) the EPICS design is that you check datatype to see which is valid.
I can't find it documented anywhere and I sent the question to Glenn. 
Once I know the code, I can make the variant.  My preference would be
to still include the other variables (because it is dumping the database content), 
with a warning note, but if you feel strongly
they should be removed, then I can.  But this is on hold for the moment.

> * The comment about "EpicsVar constructor should set the object state in a consistent way"

I removed the setters.  Are there other issues?

> * The use of "now" as a time reference that I think is contrary to Offline philosopy - what do @kutschke and @brownd1978 think about  this?

This method, asking for "n days ago" was a request from Glenn.  I think what it means, and the fact it includes "now" and can change is implicitly obvious.  But I don't feel strongly and will do whatever the consensus is. You'll see I moved it to its own CLI switch to make its meaning more clear.

FNALbuild commented:
:memo: The HEAD of `main` has changed to ab5350b042e8b50b16119c003658a7abd59dd6e1. Tests are now out of date.

rlcee commented:
I added the variant, and a warning for the typed accessors.  All three issues now waiting on input.

gaponenko commented:
Hi Ray,

On Mon, 16 May 2022, Ray Culbertson wrote:

> > * The use of "now" as a time reference that I think is contrary to Offline philosopy - what do @kutschke and @brownd1978 think about  this?
> 
> This method, asking for "n days ago" was a request from Glenn.  I think what it means, and the fact it includes "now" and can change is implicitly obvious.  But I don't feel strongly and will do whatever the consensus is. You'll see I moved it to its own CLI switch to make its meaning more clear.

While this interface "invites" creation of non-reproducible jobs,
nothing prevents a user from calling time() from their module anyway.
I am dropping my objection here, we can leave this as is.

Andrei

gaponenko commented on DbService/inc/EpicsVar.hh:
```diff
@@ -0,0 +1,63 @@
+ public:
+  typedef std::vector<EpicsVar> EpicsVec;
+
+  EpicsVar() {}
```
If the default EpicsVar constructor is not removed, it should at least set data members to deterministic values.

gaponenko commented:
Hi Ray,

On Mon, 16 May 2022, Ray Culbertson wrote:

> I added the variant

I see.  But the code does not make sense to me.  And it does not add
any safety to the num_val() etc EpicsVar methods.   What I want to do
is remove the 3 "possible types" members _num_val, _float_val, and _str_val.
Leave just the _value.   And implement the accessors like

int num_val() const { return std::get<int>(_value); }

Also in the EpicsVar constructor require that exactly two of sv[5],
sv[6], and sv[7] are "None" and fill _value from the remaining type.
Throw is this assumption does not hold.   Then we will have a guarantee
that an instance of EpicsVar holds a single value of a single type,
and that a user won't get a wrong one by mistake.

Andrei


>  , and a warning for the typed accessors.  All three issues now waiting on input.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_787-23issuecomment-2D1128125350&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=37cel5gkPi6eN2JLZmXfYXubTT4Q6UnQsq-AOaKGg7DURp3BZ85FjZfc3EwNvnwi&s=Z0vigPbJjYGlXHkGWa1va2tu_XmHQytmNvufVs5iGas&e= 
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

rlcee commented:
>int num_val() const { return std::get<int>(_value); }

I exposed the variant so the user could check it,
and avoid the exception.  This gets me thinking about
how do we actually want this to behave.  Here
are some failure patterns and thoughts on behavior
1) corruption in transfer from database.
This is caught by return codes and incorrect number 
of csv entries on a line.  Since this is a system
failure and transient, it should throw. This is what is does now.
2) variable changes from one type to another
  (this shouldn't happen, but only by policy)
3) inconsistent content in the 3 variable types
3a) number of null variables != 2
3b) string content can't be converted to variable (int and float)

Type 2) should presented to the user to deal with.
I think it is reasonable to throw on the first incorrect
access, but then the user should be given a way to 
detect the condition and use it.

Failures of type 3 might be transient corruption,
but I don't know how to separate that from the case of
unexpected content in the database, so we have to treat them the same.
These failures may be present in only a few entries
and this code should be able to effectively skip over them and not throw.
The options are to not return the measurement to the user or 
to present flags to allow the user to see and handle. I think I prefer
the latter, so the user would be aware.
We can also print warnings, but not throw.

If this analysis is OK, then to implement it, maybe there
are three ways.  Note since none of the three might be useable in type 3
failure, if you use variant, you have to include std::monostate
as as type, so it can be "empty".
1) contain all info in variant, present the variant
to the user. It throws on incorrect access, then the user
rewrites code
if (std::holds_alternative<int>(var.value())) {
  ..
} else if (std::holds_alternative<float>(var.value()))
or std::get_if, etc

2) use a variant, but hide it in accessors. If user get
a throw, they rewrite it as
if(var.isInt()) {
  var.num_val() ..
} else(var.isDouble())

3) same as 2, but without variant, just bools and PODs
I prefer this as most of us don't know variants off the top
of our heads.  But between 2 and 3, I don't feel strongly,
so I will do whatever the consensus is.

Finally, there doesn't seem to be simple way to
catch 3b) except by try/catch.  Are people OK with that?

kutschke commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 792ac91bc3f409db97296373dc8f5840b404ff74: build  (Build queue is empty)

FNALbuild commented:
:sunny: The build tests passed at 792ac91bc3f409db97296373dc8f5840b404ff74.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged 792ac91bc3f409db97296373dc8f5840b404ff74 at dee655f8fca0e0a429221ac2a155e35190f3c48b |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/scons.log). Build time: 10 min 49 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :white_check_mark: | [TODO (0) FIXME (0) in 7 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/artifact/clang-tidy.log) |


N.B. These results were obtained from a build of this Pull Request at 792ac91bc3f409db97296373dc8f5840b404ff74 after being merged into the base branch at dee655f8fca0e0a429221ac2a155e35190f3c48b.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1304/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

gaponenko commented:
Hi Ray,
   
On Tue, 17 May 2022, Ray Culbertson wrote:

> >int num_val() const { return std::get<int>(_value); }
> 
> I exposed the variant so the user could check it,
> and avoid the exception. 

I think it is good to expose the variant itself via the value()
accessor as you do.  Having num_val() etc. is just an added
convenience.

One more thought: the _datatype member is also redundant, because this
information is available in _value.  I prefer to avoid storing
redundant info, this design avoid objects that are internally
inconsistent without relying on extra implementation logic.  Can you
re-implement the datatype() accessor using the variant and not the
extra data member?


> This gets me thinking about
> how do we actually want this to behave.  Here
> are some failure patterns and thoughts on behavior
> 1) corruption in transfer from database.
> This is caught by return codes and incorrect number 
> of csv entries on a line.  Since this is a system
> failure and transient, it should throw. This is what is does now.
> 2) variable changes from one type to another
>   (this shouldn't happen, but only by policy)
> 3) inconsistent content in the 3 variable types
> 3a) number of null variables != 2
> 3b) string content can't be converted to variable (int and float)
> 
> Type 2) should presented to the user to deal with.
> I think it is reasonable to throw on the first incorrect
> access, but then the user should be given a way to 
> detect the condition and use it.
> 
> Failures of type 3 might be transient corruption,
> but I don't know how to separate that from the case of
> unexpected content in the database, so we have to treat them the same.
> These failures may be present in only a few entries

We may or may not ever see those "technical" failures.  My bigger
worry is someone cut-and-pasting a block of code and forgetting to
change float_val() to num_val(), something like that.


> and this code should be able to effectively skip over them and not throw.

I do not agree.  I think we should start by throwing an exception on
any anomaly.  Most of the hypothetical errors will not occur, and it
is good to know that this is the case.  If we silence errors
preemptively we won't know whether they occur.  They are supposedly
rare and will only be seen in production, not in tests.

Once we start getting exceptions in real use, we can analyze them and
decide how to address the problem.  "Skipping over" may not be the
best choice.  And if sometimes it is, we will just catch the thrown
exeptions for those cases.  This should be rare enough so that the
overhead is negligible.  If exeptions are not rare, there is something
else to be fixed.


> The options are to not return the measurement to the user or 
> to present flags to allow the user to see and handle. I think I prefer
> the latter, so the user would be aware.
> We can also print warnings, but not throw.
> 
> If this analysis is OK, then to implement it, maybe there
> are three ways.  Note since none of the three might be useable in type 3
> failure, 

Cleaning up the database so it does not return wrong data would be my
preferred way to address all the non-transient type 3 errors.


> if you use variant, you have to include std::monostate
> as as type, so it can be "empty".

Why would one want to have an "emtpy" EpicsVar?  This is kicking the
can down the road, forcing every user to deal with it, instead of
fixing it once and for all.

My view is that a class (like EpicsVar) should check its inputs for
all kinds of errors and throw an exeption - no "garbage in, garbage
out".  So that once I get an instance, I know the object is self
consistent and can be used without further checks.  IF we come to a
point where it is not practical to fix all the inputs and we decide to
"skip over" some entries, that skipping should be done at the level of
code that creates EpicsVar instances, not by EpicsVar itself.

Comments?
Andrei

rlcee commented:
> worry is someone cut-and-pasting a block of code and forgetting to
> change float_val() to num_val(), something like that.
In all my recent proposals, the user could never get a float value if there
was no valid float variable in a valid record, accessing the invalid float would always throw.
So I think we agree here.

The remainder of the discussion seem to come down to one point.
My understanding is that you always want to always throw on an invalid record.
I want to deliver all the database records to the user, giving them 
the tools to detect, and handle at their discretion, an ill-formed record.
I took this approach because I don't want a bad record at position N to 
block a dump of records from N-100 to N+100.  I'm thinking of someone
investigating something that happened last night, and not really having the
time to initiate an archiver failure and repair investigation.  Also I don't know 
if the archiver might regularly create an invalid record if an IOC disconnects, for example.
The IOC might even report that it is in an invalid state.
But given the length of this discussion, if you reply once more 
standing by your position, I will implement your position, and this 
can be addressed further after we get more experience.

kutschke commented:
> The remainder of the discussion seem to come down to one point. My understanding is that you always want to always throw on an invalid record. I want to deliver all the database records to the user, giving them the tools to detect, and handle at their discretion, an ill-formed record. I took this approach because I don't want a bad record at position N to block a dump of records from N-100 to N+100. I'm thinking of someone investigating something that happened last night, and not really having the time to initiate an archiver failure and repair investigation. Also I don't know if the archiver might regularly create an invalid record if an IOC disconnects, for example. The IOC might even report that it is in an invalid state. But given the length of this discussion, if you reply once more standing by your position, I will implement your position, and this can be addressed further after we get more experience.

Andrei: remember that this code will need to run in the trigger where throwing an exception can be the wrong answer.  For that use case, there must be a way for code to decide that they can survive certain classes of errors.  So they need the tools learn what the issue is and to make the decision.

For Offline, I agree that the right answer is to throw and exception.

Ray: can this be run time configurable?  Or build-time configurable?

kutschke commented:
Sorry - I replied in a hurry and forgot some of the context.  Thanks for the correction Ray.

I think that going with Andrei's approach is right.

kutschke commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 72afe45665ef53b237c675d83b036655b83b035e: build  (Build queue is empty)

FNALbuild commented:
:sunny: The build tests passed at 72afe45665ef53b237c675d83b036655b83b035e.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged 72afe45665ef53b237c675d83b036655b83b035e at dee655f8fca0e0a429221ac2a155e35190f3c48b |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/scons.log). Build time: 12 min 22 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :white_check_mark: | [TODO (0) FIXME (0) in 6 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/artifact/clang-tidy.log) |


N.B. These results were obtained from a build of this Pull Request at 72afe45665ef53b237c675d83b036655b83b035e after being merged into the base branch at dee655f8fca0e0a429221ac2a155e35190f3c48b.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1305/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

gaponenko commented on DbService/inc/EpicsVar.hh:
```diff
@@ -16,50 +16,46 @@ class EpicsVar {
-  std::string const& str_val() const { return _str_val; }
-  // this char says what the PV type is
-  char datatype() const { return _datatype; }
+  // for convenience, testing and querying variant
+  bool isInt() const { return std::holds_alternative<int>(_value); }
+  int getInt() const { return std::get<int>(_value); }
+  bool isDouble() const { return std::holds_alternative<double>(_value); }
+  double getDouble() const { return std::get<double>(_value); }
+  bool isString() const { return std::holds_alternative<std::string>(_value); }
+  std::string getString() const { return std::get<std::string>(_value); }
```
Nice!

gaponenko reviewed (approved):
Ray, thank you for making the changes!
Andrei

FNALbuild commented:
:hourglass: The following tests have been triggered for f57dd2231b6eb984bfbeeebfd0aa124ff5cff12b: build  (Build queue is empty)

FNALbuild commented:
:sunny: The build tests passed at f57dd2231b6eb984bfbeeebfd0aa124ff5cff12b.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged f57dd2231b6eb984bfbeeebfd0aa124ff5cff12b at dee655f8fca0e0a429221ac2a155e35190f3c48b |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/scons.log). Build time: 12 min 10 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :white_check_mark: | [TODO (0) FIXME (0) in 8 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/artifact/clang-tidy.log) |


N.B. These results were obtained from a build of this Pull Request at f57dd2231b6eb984bfbeeebfd0aa124ff5cff12b after being merged into the base branch at dee655f8fca0e0a429221ac2a155e35190f3c48b.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1306/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

brownd1978 commented:
I've been lurking on this thread.  I'm happy with the new design using
(exposed) std::variant, that's the best way to deal with this ambiguity
I agree with Andrei, throwing an exception is the best strategy for dealing
with unexpected data.  If this code is ever run in a critical environment
where termination is
unacceptable we just need to catch those throws appropriately.
If it turns out the database or epics are regularly sending corrupt data
then we have bigger problems than this.
Dave

On Tue, May 17, 2022 at 11:23 PM Andrei Gaponenko ***@***.***>
wrote:

> Hi Ray,
>
> On Tue, 17 May 2022, Ray Culbertson wrote:
>
> > >int num_val() const { return std::get<int>(_value); }
> >
> > I exposed the variant so the user could check it,
> > and avoid the exception.
>
> I think it is good to expose the variant itself via the value()
> accessor as you do. Having num_val() etc. is just an added
> convenience.
>
> One more thought: the _datatype member is also redundant, because this
> information is available in _value. I prefer to avoid storing
> redundant info, this design avoid objects that are internally
> inconsistent without relying on extra implementation logic. Can you
> re-implement the datatype() accessor using the variant and not the
> extra data member?
>
>
> > This gets me thinking about
> > how do we actually want this to behave. Here
> > are some failure patterns and thoughts on behavior
> > 1) corruption in transfer from database.
> > This is caught by return codes and incorrect number
> > of csv entries on a line. Since this is a system
> > failure and transient, it should throw. This is what is does now.
> > 2) variable changes from one type to another
> > (this shouldn't happen, but only by policy)
> > 3) inconsistent content in the 3 variable types
> > 3a) number of null variables != 2
> > 3b) string content can't be converted to variable (int and float)
> >
> > Type 2) should presented to the user to deal with.
> > I think it is reasonable to throw on the first incorrect
> > access, but then the user should be given a way to
> > detect the condition and use it.
> >
> > Failures of type 3 might be transient corruption,
> > but I don't know how to separate that from the case of
> > unexpected content in the database, so we have to treat them the same.
> > These failures may be present in only a few entries
>
> We may or may not ever see those "technical" failures. My bigger
> worry is someone cut-and-pasting a block of code and forgetting to
> change float_val() to num_val(), something like that.
>
>
> > and this code should be able to effectively skip over them and not throw.
>
> I do not agree. I think we should start by throwing an exception on
> any anomaly. Most of the hypothetical errors will not occur, and it
> is good to know that this is the case. If we silence errors
> preemptively we won't know whether they occur. They are supposedly
> rare and will only be seen in production, not in tests.
>
> Once we start getting exceptions in real use, we can analyze them and
> decide how to address the problem. "Skipping over" may not be the
> best choice. And if sometimes it is, we will just catch the thrown
> exeptions for those cases. This should be rare enough so that the
> overhead is negligible. If exeptions are not rare, there is something
> else to be fixed.
>
>
> > The options are to not return the measurement to the user or
> > to present flags to allow the user to see and handle. I think I prefer
> > the latter, so the user would be aware.
> > We can also print warnings, but not throw.
> >
> > If this analysis is OK, then to implement it, maybe there
> > are three ways. Note since none of the three might be useable in type 3
> > failure,
>
> Cleaning up the database so it does not return wrong data would be my
> preferred way to address all the non-transient type 3 errors.
>
>
> > if you use variant, you have to include std::monostate
> > as as type, so it can be "empty".
>
> Why would one want to have an "emtpy" EpicsVar? This is kicking the
> can down the road, forcing every user to deal with it, instead of
> fixing it once and for all.
>
> My view is that a class (like EpicsVar) should check its inputs for
> all kinds of errors and throw an exeption - no "garbage in, garbage
> out". So that once I get an instance, I know the object is self
> consistent and can be used without further checks. IF we come to a
> point where it is not practical to fix all the inputs and we decide to
> "skip over" some entries, that skipping should be done at the level of
> code that creates EpicsVar instances, not by EpicsVar itself.
>
> Comments?
> Andrei
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/787#issuecomment-1129525642>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH574E6GWBC54EIIB2HT3VKRPB5ANCNFSM5VJJGZMQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
David Brown            ***@***.***
Office Phone (510) 486-7261
Lawrence Berkeley National Lab
M/S 50R5008 (50-6026C)  Berkeley, CA 94720
